src/api/server.rkt

```diff
@@ -53,75 +82,394 @@
+  (define tree
+    (let loop ([expr expr] [err local-error])
+      (match expr
+        [(list op args ...)
+         ;; err => (List (listof Integer) List ...)
+         (hasheq 'e
+                 (~a op)
+                 'avg-error
+                 (format-bits (errors-score (first err)))
+                 'children
+                 (map loop args (rest err)))]
+        ;; err => (List (listof Integer))
+        [_ (hasheq 'e (~a expr) 'avg-error (format-bits (errors-score (first err))) 'children '())])))
+  (hasheq 'command (get-command herbie-result) 'tree tree 'job job-id 'path (make-path job-id)))
 
-; Encapsulates semaphores and async part of jobs.
-(define (start-work job)
-  (define job-id (compute-job-id job))
-  (hash-set! *job-status* job-id (*timeline*))
-  (define sema (make-semaphore))
-  (hash-set! *job-sema* job-id sema)
-  (thread-send *worker-thread* (work job-id job sema))
-  job-id)
+(define (make-sample-result herbie-result test job-id)
+  (define pctx (job-result-backend herbie-result))
+  (define repr (context-repr (test-context test)))
+  (hasheq 'command
+          (get-command herbie-result)
+          'points
+          (pcontext->json pctx repr)
+          'job
+          job-id
+          'path
+          (make-path job-id)))
+
+(define (make-calculate-result herbie-result job-id)
+  (hasheq 'command
+          (get-command herbie-result)
+          'points
+          (job-result-backend herbie-result)
+          'job
+          job-id
+          'path
+          (make-path job-id)))
+
+(define (make-cost-result herbie-result job-id)
+  (hasheq 'command
+          (get-command herbie-result)
+          'cost
+          (job-result-backend herbie-result)
+          'job
+          job-id
+          'path
+          (make-path job-id)))
+
+(define (make-error-result herbie-result job-id)
+  (define errs
+    (for/list ([pt&err (job-result-backend herbie-result)])
+      (define pt (first pt&err))
+      (define err (second pt&err))
+      (list pt (format-bits (ulps->bits err)))))
+  (hasheq 'command (get-command herbie-result) 'points errs 'job job-id 'path (make-path job-id)))
+
+(define (make-exacts-result herbie-result job-id)
+  (hasheq 'command
+          (get-command herbie-result)
+          'points
+          (job-result-backend herbie-result)
+          'job
+          job-id
+          'path
+          (make-path job-id)))
+
+(define (make-improve-result herbie-result test job-id)
+  (define ctx (ctx-hash-table (test-context test)))
+  (define backend (job-result-backend herbie-result))
+  (define job-time (job-result-time herbie-result))
+  (define warnings (job-result-warnings herbie-result))
+  (define timeline (job-result-timeline herbie-result))
+
+  (define repr (test-output-repr test))
+  (define backend-hash #f)
+
+  (match (job-result-status herbie-result)
+    ['success (set! backend-hash (backend-improve-result-hash-table backend repr test))]
+    [else (set! backend-hash #f)])
```

Comment:
Also `else` here isn't a keyword, it just binds the variable name "else". Maybe that's fine? But I usually use `_`, which is the pattern that always matches.